lms/djangoapps/lti_provider/models.py

```diff
@@ -25,25 +30,69 @@ class LtiConsumer(models.Model):
     consumer_secret = models.CharField(max_length=32, unique=True)
 
 
+class OutcomeService(models.Model):
+    """
+    Model for a single outcome service associated with an LTI consumer. Note
+    that a given consumer may have more than one outcome service URL over its
+    lifetime, so we need to store the outcome service separately from the
+    LtiConsumer model.
+
+    An outcome service can be identified in two ways, depending on the
+    information provided by an LTI launch. The ideal way to identify the service
+    is by instance_guid, which should uniquely identify a consumer. However that
+    field is optional in the LTI launch, and so if it is missing we can fall
+    back on the consumer key (which should be created uniquely for each consumer
+    although we don't have a technical way to guarantee that).
+    """
+    lis_outcome_service_url = models.CharField(max_length=255)
+    instance_guid = models.CharField(max_length=255, null=True)
+    consumer_key = models.CharField(max_length=32, db_index=True)
+
+
+class GradedAssignment(models.Model):
+    """
+    Model representing a single launch of a graded assignment by an individual
+    user. There will be a row created here only if the LTI consumer may require
+    a result to be returned from the LTI launch (determined by the presence of
+    the lis_result_sourcedid parameter in the launch POST). There will be only
+    one row created for a given user/course/usage/consumer combination; repeated
+    launches of the same content by the same user from the same LTI consumer
+    will not add new rows to the table.
+    """
+    user = models.ForeignKey(User, db_index=True)
+    course_key = CourseKeyField(max_length=255, db_index=True)
+    usage_key = UsageKeyField(max_length=255, db_index=True)
+    outcome_service = models.ForeignKey(OutcomeService)
+    lis_result_sourcedid = models.CharField(max_length=255, db_index=True, unique=True)
```

Comment:
I was wary about adding a constraint (I had one in for a while, but took it out later). A close reading of the LTI 1.1 spec says that a given lis_result_sourcedid refers to a single row in a gradebook, which would imply that a unique result ID/outcome service should map to a unique user/course/usage combination. There is an edge case, though, where you could add the same edX content twice to the same campus course, which should legitimately create two records with the same user/course_key/usage_key/outcome_service values. Since it's an unlikely case I've got adding the logic to handle it on my todo list rather than in this change, but I didn't want to add constraints to the database that we'd have to take out.

Of course, while writing that it occurs that the unique constraint on lis_result_sourcedid is wrong - it should be unique_together over lis_result_sourcedid and outcome_service. I'll make that change.